subsys/bluetooth/mesh/gatt_cli.c

```diff
@@ -0,0 +1,374 @@
+
+		server->discover.uuid = &server->gatt->data_out_uuid.uuid;
+		server->discover.start_handle = server->svc_start_handle + 1;
+		server->discover.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+
+		err = bt_gatt_discover(conn, &server->discover);
+		if (err) {
+			BT_DBG("Discover GATT data out char failed (err %d)", err);
+		}
+	} else if (!bt_uuid_cmp(server->discover.uuid,
+				&server->gatt->data_out_uuid.uuid)) {
+		server->discover.uuid = &server->gatt->data_out_cccd_uuid.uuid;
+		server->discover.start_handle = attr->handle + 2;
+		server->discover.type = BT_GATT_DISCOVER_DESCRIPTOR;
+
+		err = bt_gatt_discover(conn, &server->discover);
+		if (err) {
+			BT_DBG("Discover GATT CCCD failed (err %d)", err);
+		}
+	} else {
+		server->subscribe.notify = notify_func;
+		server->subscribe.value = BT_GATT_CCC_NOTIFY;
+		server->subscribe.ccc_handle = attr->handle;
+		server->subscribe.value_handle = attr->handle - 1;
+
+		err = bt_gatt_subscribe(conn, &server->subscribe);
+		if (err && err != -EALREADY) {
+			BT_DBG("Subscribe failed (err %d)", err);
+		} else {
+			BT_DBG("[SUBSCRIBED]");
+
+			server->gatt->link_open(conn);
+		}
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+int bt_mesh_gatt_send(struct bt_conn *conn,
+		      const void *data, uint16_t len,
+		      bt_gatt_complete_func_t end, void *user_data)
+{
+	struct bt_mesh_gatt_server *server;
+
+	server = find_server(conn);
+	if (!server) {
+		return -ENOTCONN;
+	}
+
+	BT_DBG("%u bytes: %s", len, bt_hex(data, len));
+
+	return bt_gatt_write_without_response_cb(conn, server->data_in_handle,
+						 data, len, false, end, user_data);
+}
+
+static void gatt_connected(struct bt_conn *conn, uint8_t conn_err)
+{
+	struct bt_mesh_gatt_server *server;
+	struct bt_conn_info info;
+	int err;
+
+	bt_conn_get_info(conn, &info);
+	if (info.role != BT_CONN_ROLE_CENTRAL) {
+		return;
+	}
+
+	server = find_server(conn);
+	if (!server) {
+		return;
+	}
+
+	if (conn_err) {
+		BT_ERR("Failed to connect GATT Services(%u)", conn_err);
+
+		bt_conn_unref(server->conn);
+
+		server->conn = NULL;
+		(void)bt_mesh_scan_enable();
+
+		return;
+	}
+
+	BT_DBG("conn %p err 0x%02x", (void *)conn, conn_err);
+
+	server->gatt->connected(conn, server->user_data);
+
+	(void)bt_mesh_scan_enable();
+
+	server->discover.uuid = &server->gatt->srv_uuid.uuid;
+	server->discover.func = discover_func;
+	server->discover.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE;
+	server->discover.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+	server->discover.type = BT_GATT_DISCOVER_PRIMARY;
+	err = bt_gatt_discover(server->conn, &server->discover);
+	if (err) {
+		BT_ERR("Unable discover GATT Services (err %d)", err);
+	}
+}
+
+static void gatt_disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	struct bt_conn_info info;
+	struct bt_mesh_gatt_server *server;
+
+	bt_conn_get_info(conn, &info);
+	if (info.role != BT_CONN_ROLE_CENTRAL) {
+		return;
+	}
+
+	server = find_server(conn);
+	if (!server) {
+		return;
+	}
```

Comment:
And here (there are many more places - I won't bother commenting on each one separately)